### PR TITLE
Restructure matrix_free/cuda_tensor_product_kernels.h [WIP]

### PR DESCRIPTION
**Pull request by peterrum**
This involves:
- splitting up `EvaluatorTensorProduct` into `EvaluatorTensorProduct` and `FEEvaluationImpl`
- introduction of `FEEvaluationCollocationImpl`
- ~~specialize `EvaluatorTensorProduct` for `EvaluatorVariant::evaluate_evenodd`~~

Please not that the first two commits are part of other PRs.

### CODE REVIEW

**Comment by peterrum** [comment]
@Rombur To test this, I can add step-64 using instead of `const QGauss<1> quad(fe_degree + 1)` `const QGaussLobatto<1> quad(fe_degree + 1)`. Although, I am not quite sure how I will generate the output, since the build-system is not working properly on the GPU cluster I have access to...

**Comment by masterleinad** [comment]
> @Rombur To test this, I can add step-64 using instead of `const QGauss<1> quad(fe_degree + 1)` `const QGaussLobatto<1> quad(fe_degree + 1)`. Although, I am not quite sure how I will generate the output, since the build-system is not working properly on the GPU cluster I have access to...

We can test the pull requests and generate output. Just saying that examples are not good as tests if they are not converted to something appropriate for the `tests` folder if that is what you had in mind.

File: include/deal.II/matrix_free/cuda_tensor_product_kernels.h [hunk 1]
```diff
@@ -90,22 +90,54 @@ namespace CUDAWrappers
...
-      gradients(Number shape_gradients[], const Number *in, Number *out) const;
+      gradients(const Number *in, Number *out) const;
+
+      /**
+       * Evaluate the gradient of a finite element function at the quadrature
+       * points for a given @p direction.
+       */
+      template <int direction, bool dof_to_quad, bool add, bool in_place>
+      __device__ void
+      co_gradients(const Number *in, Number *out) const;
```
Comment (Rombur) [inline include/deal.II/matrix_free/cuda_tensor_product_kernels.h]: The comment is exactly the same for `co_gradients` and `gradients`

File: include/deal.II/matrix_free/cuda_tensor_product_kernels.h [hunk 2]
```diff
@@ -90,22 +90,54 @@ namespace CUDAWrappers
...
+      values(const Number *in, Number *out) const;
 
       /**
        * Evaluate the gradient of a finite element function at the quadrature
        * points for a given @p direction.
        */
       template <int direction, bool dof_to_quad, bool add, bool in_place>
       __device__ void
-      gradients(Number shape_gradients[], const Number *in, Number *out) const;
+      gradients(const Number *in, Number *out) const;
```
Comment (Rombur) [inline include/deal.II/matrix_free/cuda_tensor_product_kernels.h]: So we are reverting the interface that you just introduced?

Comment (peterrum) [reply include/deal.II/matrix_free/cuda_tensor_product_kernels.h]: I wanted to save a reference to the right `shape_value` and `shape_gradient` as a class member, but that did not work. Do you know if you can do that for `__constant__` variables?

Comment (peterrum) [reply include/deal.II/matrix_free/cuda_tensor_product_kernels.h]: In the case of `MatrixFree` (actual), the`shape_values` and `shape_gradients` are passed by reference in the constructor:
```cpp
    EvaluatorTensorProduct(const AlignedVector<Number2> &shape_values,
                           const AlignedVector<Number2> &shape_gradients,
                           const AlignedVector<Number2> &shape_hessians,
                           const unsigned int            dummy1 = 0,
                           const unsigned int            dummy2 = 0)
      : shape_values(shape_values.begin())
      , shape_gradients(shape_gradients.begin())
      , shape_hessians(shape_hessians.begin())
    {}
```

**Comment by peterrum** [comment]
Since I don't plan to work on this PR anytime soon, let me close this one!

**Pull request closed without merging**